sql/logictest: TestLogic failed

[(sql/logictest).TestLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2550107&tab=buildLog) on [release-20.2@e60df4714f55d37f2e0d680a9449946e8d40fa41](https://github.com/cockroachdb/cockroach/commits/e60df4714f55d37f2e0d680a9449946e8d40fa41):

```
=== RUN   TestLogic
--- FAIL: TestLogic (1045.02s)
    logic.go:2819: randomize coldata.BatchSize to 2
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic220642731
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestLogic/local/schema_change_in_txn/create_as_drop_and_create_in_txn
--- progress: testdata/logic_test/schema: 59 statements/queries
--- progress: testdata/logic_test/rename_view: 41 statements/queries
--- done: testdata/logic_test/rename_view with config local: 42 tests, 0 failures
            --- FAIL: TestLogic/local/schema_change_in_txn/create_as_drop_and_create_in_txn (0.35s)
                logic.go:2357: 
                    
                    testdata/logic_test/schema_change_in_txn:403: 
                    expected success, but found
                    (40001) restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=875868d5 key=/Table/SystemConfigSpan/Start pri=0.00510709 epo=0 ts=1609309009.872583679,1 min=1609309009.732064209,0 seq=39} lock=true stat=PENDING rts=1609309009.732064209,0 wto=false max=1609309010.232064209,0
                logic.go:1670: 
                     pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=875868d5 key=/Table/SystemConfigSpan/Start pri=0.00510709 epo=0 ts=1609309009.872583679,1 min=1609309009.732064209,0 seq=39} lock=true stat=PENDING rts=1609309009.732064209,0 wto=false max=1609309010.232064209,0
=== RUN   TestLogic/local/schema_change_in_txn
=== PAUSE TestLogic/local/schema_change_in_txn
=== CONT  TestLogic/local/schema_change_in_txn
        --- FAIL: TestLogic/local/schema_change_in_txn (10.63s)
=== RUN   TestLogic/local
    --- FAIL: TestLogic/local (0.64s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

Related:
- #58269 sql/logictest: TestLogic failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1)

- #56551 sql/logictest: TestLogic failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1093,6 +1093,12 @@ type logicTest struct {
 	// with 25% probability, a random value in [2, 100] range with 25%
 	// probability, or default max batch size with 50% probability.
 	randomizedMutationsMaxBatchSize int
+
+	// skipOnRetry is explicitly set to true by the skip_on_retry directive.
+	// If true, serializability violations in the test when unexpected cause the
+	// entire test to be skipped and the below skippedOnRetry to be set to true.
+	skipOnRetry    bool
+	skippedOnRetry bool
 }
 
 func (t *logicTest) t() *testing.T {
@@ -1679,6 +1685,7 @@ func (t *logicTest) processTestFile(path string, config testClusterConfig) error
 					t.Error(err)
 				}
 			})
+			t.maybeSkipOnRetry(nil)
 		}
 	}
 
@@ -1796,6 +1803,8 @@ func (t *logicTest) processSubtest(
 				)
 			}
 			repeat = count
+		case "skip_on_retry":
+			t.skipOnRetry = true
 
 		case "sleep":
 			var err error
@@ -2251,6 +2260,17 @@ func (t *logicTest) processSubtest(
 	return s.Err()
 }
 
+// Some tests encounter serializability failures sometimes and indicate
+// that they want to just get skipped when that happens.
+func (t *logicTest) maybeSkipOnRetry(err error) {
+	if pqErr := (*pq.Error)(nil); t.skippedOnRetry ||
+		(t.skipOnRetry && errors.As(err, &pqErr) &&
+			pgcode.MakeCode(string(pqErr.Code)) == pgcode.SerializationFailure) {
+		t.skippedOnRetry = true
+		skip.WithIssue(t.t(), 53724)
+	}
+}
+
 // verifyError checks that either no error was found where none was
 // expected, or that an error was found when one was expected.
 // Returns a nil error to indicate the behavior was as expected.  If
@@ -2261,6 +2281,7 @@ func (t *logicTest) verifyError(
 	sql, pos, expectErr, expectErrCode string, err error,
 ) (bool, error) {
 	if expectErr == "" && expectErrCode == "" && err != nil {
+		t.maybeSkipOnRetry(err)
 		cont := t.unexpectedError(sql, pos, err)
 		if cont {
 			// unexpectedError() already reported via t.Errorf. no need for more.

--- a/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn
+++ b/pkg/sql/logictest/testdata/logic_test/schema_change_in_txn
@@ -3,6 +3,11 @@
 statement ok
 SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false
 
+
+# Skip the rest of the test if a retry occurs. They can happen and are fine
+# but there's no way to encapsulate that in logictests.
+skip_on_retry
+
 subtest create_and_add_fk_in_same_txn
 
 statement ok

--- a/pkg/sql/logictest/testdata/logic_test/skip_on_retry
+++ b/pkg/sql/logictest/testdata/logic_test/skip_on_retry
@@ -1,0 +1,24 @@
+# LogicTest: local
+
+# This test demonstrates that the skip_on_retry directive works as expected.
+# The directive applies to the entire remainder of the file.
+
+statement ok
+BEGIN; SELECT * FROM system.namespace LIMIT 1;
+
+statement error pgcode 40001 TransactionRetryWithProtoRefreshError
+SELECT crdb_internal.force_retry('1h');
+
+statement ok
+ROLLBACK;
+
+skip_on_retry
+
+subtest skip_this_subtest
+
+statement ok
+BEGIN; SELECT * FROM system.namespace LIMIT 1;
+
+statement ok
+SELECT crdb_internal.force_retry('1h');
+

